### PR TITLE
apps/postgresql: Receiver Enhancements (Metric Additions)

### PR DESCRIPTION
**Pull request by schmikei**
## Description
- Updates operations-collector, this includes a go1.17 => go1.18 update
- Postgres enhancements are currently behind a featuregate `receiver.postgresql.emitMetricsWithResourceAttributes` for more information on transitioning, please reference https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/postgresqlreceiver#transition-schedule
- The postgresql receiver has been enhanced to take resource attributes now so we must transform metrics to fit old style

## How has this been tested?
Integration testing and updates to confgenerator unit testing

## Checklist:
- Unit tests
  - [x] Unit tests do not apply.
  - [ ] Unit tests have been added/modified and passed for this PR.
- Integration tests
  - [ ] Integration tests do not apply.
  - [x] Integration tests have been added/modified and passed for this PR.
- Documentation
  - [ ] This PR introduces no user visible changes.
  - [x] This PR introduces user visible changes and the corresponding documentation change has been made.
- Minor version bump
  - [ ] This PR introduces no new features.
  - [ ] This PR introduces new features, and there is a separate PR to bump the [minor version](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/VERSION) since the last [release](https://github.com/GoogleCloudPlatform/ops-agent/releases) already.

### CODE REVIEW

File: cmd/ops_agent_windows/main_windows.go
Comment (schmikei) [inline cmd/ops_agent_windows/main_windows.go:112]: @qingling128 @ridwanmsharif 

So this feature gate is disabled by default will be enabled by default by the end of September hopefully in Otel. I was just double-checking with you guys on whether this was safe to add temporarily or will it cause issues down the line when the feature gate is removed?

Comment (schmikei) [reply cmd/ops_agent_windows/main_windows.go:112]: Tested an upgrade and figured I'd share what I've learned:

1. Service files stay in place even in upgrades.
2. Collector fails to start if a featuregate is specified that does not exist which in eventuality will become the case so this would eventually become an issue as users would have to modify their service file even if they don't use the postgresqlreceiver.

**Comment by schmikei** [comment]
image

File: cmd/ops_agent_windows/main_windows.go
Comment (ridwanmsharif) [reply cmd/ops_agent_windows/main_windows.go:112]: Hmmmmm, it feels like to me that waiting for v0.60.0 is more reasonable. If we wanted to release this earlier than the september release date, we could potentially see if we can update the packaging/upgrade scripts to pick up the changes to the service unit files too. 

cc @qingling128 for opinions

Comment (qingling128) [reply cmd/ops_agent_windows/main_windows.go:112]: Do we have to remove the `receiver.postgresql.emitMetricsWithResourceAttributes` feature gate in Sept? Is it expected to be deprecated at that time? If so, +1 for waiting for `0.60.0`. If the `receiver.postgresql.emitMetricsWithResourceAttributes` feature gate is expected to be around for good, we can keep it and release now.

Comment (schmikei) [reply cmd/ops_agent_windows/main_windows.go:112]: @qingling128 

Yeah so for this case the featuregate is expected to be a temporary transition period that defaults to future behavior in september in the 0.60.0 release. Technically is deprecated now but default behavior at the current moment. The default behavior in 0.60.0 will be the future behavior and therefore doesn't need to specify the featuregate. 

I heard from @StefanKurek that @ridwanmsharif was going to look into fixing the issue with the ops agent not properly updating service info on upgrade. are you guys going to look into that while waiting for the v0.60.0 release

**Comment by schmikei** [comment]
All kokoro tests passed on commit 546d680042941dcef7e5f0aaecb482d80224d0eb

Let me know if you would like a rerun @ridwanmsharif @qingling128 

image

File: integration_test/third_party_apps_data/applications/postgresql/sles/install
Comment (igorpeshansky) [inline integration_test/third_party_apps_data/applications/postgresql/sles/install:29]: These fixes would apply to the current state as well, won't they? Can we please extract them into a separate PR?

Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/sles/install:29]: Sure thing one moment 👍

File: integration_test/third_party_apps_data/applications/postgresql/centos_rhel/install
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/centos_rhel/install:25]: The `sudo su postgres` lines seem not related to the installation. Can we move them to:

https://github.com/GoogleCloudPlatform/ops-agent/blob/master/integration_test/third_party_apps_data/applications/postgres/exercise

Similar to https://github.com/GoogleCloudPlatform/ops-agent/blob/master/integration_test/third_party_apps_data/applications/nginx/exercise, which is where commands go when they are not distro specific and the goal is to trigger some metrics / logs for testing instead of installing the application.

File: docs/postgresql.md
Comment (qingling128) [inline docs/postgresql.md:8]: The `docs` folder can probably be deleted now that we track everything in `metadata.yaml`. [Documentation](https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent/third-party) is automatically generated based on `metadata.yaml` so we don't need to manually update the `.md` files in this repo any more.

CC: @sophieyfang for confirmation

File: integration_test/third_party_apps_data/applications/postgresql/centos_rhel/install
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/centos_rhel/install:25]: Sure can do that, would the modification of the postgres user still be included in the `install` file then or moved to `exercise` from your perspective?

File: cmd/ops_agent_windows/main_windows.go
Comment (ridwanmsharif) [reply cmd/ops_agent_windows/main_windows.go]: Do we expect the v0.60.0 release to break some existing metrics for the Ops Agent? I'm wondering if upgrading to that version will take a while for us (i.e. if we have to deal with the breaking changes in the Ops Agent first).

If that is the case, we may need to prioritize the bug you identified now where the service unit files are not being updates with the upgrade.

Comment (schmikei) [reply cmd/ops_agent_windows/main_windows.go]: For postgres, v0.60.0 with the current modifications of flattening resource attributes, should be a purely additive change. Old dashboards/alerts should remain unaffected but new metrics should be added.

Comment (schmikei) [reply cmd/ops_agent_windows/main_windows.go]: This PR brings in essentially the v0.60.0 postgresqlreceiver https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/pull/120

File: docs/postgresql.md
Comment (qingling128) [reply docs/postgresql.md:8]: Confirmed with @sophieyfang that they can be deleted. Sending a separate PR to delete the whole folder. For this PR, we can just ignore this file.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:55]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) has this as `workload.googleapis.com/postgresql.bgwriter.checkpoints`

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approver double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:65]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) has this as `workload.googleapis.com/postgresql.bgwriter.maxwriten.count`

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approver double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:83]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) has both `workload.googleapis.com/postgresql.connection.open` and `workload.googleapis.com/postgresql.connection.max`

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approver double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:96]: For `workload.googleapis.com/postgresql.index.scans`, the scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) does not include any labels.

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approver double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:104]: For `workload.googleapis.com/postgresql.index.size`, the scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) does not include any labels.

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approver double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:120]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) does not have labels for `workload.googleapis.com/postgresql.replication.data_delay`

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approval double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:124]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) has this as `workload.googleapis.com/postgresql.replication.delay`

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approval double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:117]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) has this as `workload.googleapis.com/postgresql.row.operations` with labels `operation=delete|fetch|hot_update|insert|return|update`.

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approval double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:144]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) does not include labels for `workload.googleapis.com/postgresql.table.count`

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approval double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:151]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) include no labels for `workload.googleapis.com/postgresql.table.size`

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approval double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:158]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) has this as `workload.googleapis.com/postgresql.vacuum.count`

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approval double check it? If not, can we fix the translation so it matches the scope doc?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (qingling128) [inline integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:168]: The scope [doc](https://docs.google.com/document/d/1M9r8CqZl2W9fYz8brzlWntEAqJUC1K43LHGvGkbpiPs/edit) does not include a `replication_client` label.

Is this an intentional change to the scope doc? If so, can we send a suggested edit to the scope doc and have the approval double check it? If not, can we fix the translation so it matches the scope doc?

**Review by qingling128 — COMMENTED** [review]
Compared the implementation with the scope doc. There are a few discrepancies that are not unclear whether they are intentional or by accident. We can discuss them case by case. See detailed comments.

The main concern is:
* Are the newly added labels gonna cause granularity / scalability burden that is worth revisiting.
* If there were some proposals to change what metrics should look like, the suggestion should be recorded in the scope doc with reasoning.

**Comment by qingling128** [comment]
CC @braydonk for shadowing this PR review.

File: docs/postgresql.md
Comment (qingling128) [reply docs/postgresql.md]: Sent https://github.com/GoogleCloudPlatform/ops-agent/pull/842

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:65]: Intentional change, due to this https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13328#discussion_r945933195

Will make a suggestion

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:168]: This was intentional as to properly identify the metric. See https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-REPLICATION-VIEW for more information on the table.

Will send a suggestion

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:158]: Intentional change as the identifying information relay to a table. Will make a suggestion on scope doc.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:151]: It needs these labels to be properly identified per table. Will make a suggestion to the scope doc.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:117]: This was intentional as this was the pre-existing metric `postgresql.operations` was not sure it was worth breaking users to make this name change. Will make a suggestion to scope doc.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:104]: We needed to identify the index with labels for its size. Will make a suggestion to scope doc.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:96]: We needed to identify the index with labels for its size. Will make a suggestion to scope doc.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:55]: It appears that this metric name matches when I looked at the doc? Do you mind taking a second pass to make sure its not `postgresql.bgwriter.checkpoint.count`?

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:124]: Replication delay is the same value as `wal.lag`. Made a comment outlining this but there should be only one metric for this use case.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:144]: Databases give context to the table count. Made a suggestion to the scope doc.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:120]: Made a suggestion in the scope doc to resolve this as it identifies the data_delay. 👍

**Comment by schmikei** [comment]
>     * Are the newly added labels gonna cause granularity / scalability burden that is worth revisiting.

I don't believe that the newly added labels should be a large cause of concern. The most nested would be a metric datapoint with a value for 
`database, table, and index` which in most database systems only have less than 5 indexes per table, [best practices](https://www.pgmustard.com/blog/indexing-best-practices-postgresql) suggest only indexing columns that are filterable. 

>     * If there were some proposals to change what metrics should look like, the suggestion should be recorded in the scope doc with reasoning.

I submitted a bunch of suggestions based off what I found. Let me know who to ping or if further clarification is needed on them.

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (braydonk) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:55]: It looks to me like it says the otel metric name is `postgresql.bgwriter.checkpoint.count` and that we're expecting to emit `postgresql.bgwriter.checkpoints` from the Ops Agent.

Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:55]: Ahh I see I was looking at the otel vs ops agent metric now, thanks for the catch! Will make a suggestion!

File: integration_test/third_party_apps_data/applications/postgresql/metadata.yaml
Comment (ridwanmsharif) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:117]: https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent/third-party/postgresql#monitored-metrics

From the list of metrics above it doesn't seem like `postgresql.operations` was ever monitored (or atleast doesn't seem so from the public documentation and the metadata.yaml file)

This is interesting to me though because of: [b/246310954](http://b/246310954) | P2 | integration_test: test that no metric is reported for third party apps that don't have a corresponding metadata.yaml entry. 

Are we actually collecting more metrics quietly and not documenting? I looked at the old Postgres scope doc from last year and it says we will monitor the same as the legacy agent metrics: https://cloud.google.com/monitoring/api/metrics_agent#agent-postgresql

But that is a much larger list than what we claim to collect. We should double check what we're collecting and have the metadata and public docs reflect that.

Comment (schmikei) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:117]: Hmm interesting... There were metrics added it appears from november 2021 into Otel https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/09d2b3eb52d6870ecbb46dbcc29dbff88294fe27

Also appears the docs (now replaced by metadata.yaml) were updated, but this was before the metadata.yaml seemed to have a strong foothold. https://github.com/GoogleCloudPlatform/ops-agent/pull/320/files

Metrics probably wouldn't have shown up because no user table was created. In the `exercise` I create tables and indexes to get these metrics to show up.

Comment (ridwanmsharif) [reply integration_test/third_party_apps_data/applications/postgresql/metadata.yaml:117]: image

I did a very unscientific test to see what was being collected now. Looking at the integration test VMs and what they collected a few months ago, it looks like only the publicly documented metrics were actually being collected. For all other metrics (added in this PR probably), they had no data. This implies to me that we actually only collected the publicly documented ones since the integration test does run the `enable` and `exercise` scripts. But can we verify this with a manual install and run?

I suspect you're right about these metrics being collected only in some circumstances and our docs and metadata files don't capture that scenario. If that is the case, this should be fixed

**Review by braydonk — APPROVED** [review]

**Comment by ridwanmsharif** [comment]
Looks like a flake to me. Re-running to make sure.

**Comment by ridwanmsharif** [comment]
`TestExcludeLogsParseJsonOrder/windows-2012-r2` failed. Looks unrelated so merging this now.

**Pull request merged**
